src/main/java/com/jcraft/jsch/KnownHosts.java

```diff
@@ -490,18 +510,24 @@ private String deleteSubString(String hosts, String host){
     return hosts;
   }
 
-  private MAC getHMACSHA1(){
-    if(hmacsha1==null){
-      try{
-        Class<? extends MAC> c=Class.forName(JSch.getConfig("hmac-sha1")).asSubclass(MAC.class);
-        hmacsha1=c.getDeclaredConstructor().newInstance();
-      }
-      catch(Exception e){ 
-        System.err.println("hmacsha1: "+e); 
-      }
+  MAC getHMACSHA1(){
+    if (hmacsha1 == null){
+      hmacsha1 = createHMAC(JSch.getConfig("hmac-sha1"));
     }
+    
     return hmacsha1;
   }
+  
+  MAC createHMAC(String hmacClassname) {
+    try{
+      Class<? extends MAC> c=Class.forName(hmacClassname).asSubclass(MAC.class);
+      return c.getDeclaredConstructor().newInstance();
+    }
+    catch(Exception e){ 
+      jsch.getInstanceLogger().log(Logger.ERROR, "unable to instantiate HMAC-class " + hmacClassname + ", using SHA-1 as fallback", e);
+      return new HMACSHA1(); // the default in config
+    }
+  }
```

Comment:
So I've been thinking about this some more: I think instead of falling back to the `com.jcraft.jsch.jce.HMACSHA1` implementation, I think we should simply throw a `JSchException` if creating the `hmacClassname` fails.
I think this will work because:

1. The `KnownHosts` class is package private since the 0.2.1 release, so we have some API flexibility.
2. All the locations in the `JSch` and `Session` classes where we instantiate a `KnownHosts` instance all appear to be in methods that already allow throwing a `JSchException`.

I also think this makes more sense from a user perspective, since it would immediately inform a user that the class they configured as `hmac-sha1` was not usable, instead of hiding it away in a way that they may not realize (especially if they don't have a logger setup).

Reply:
I can do the change but a misconfigured hmac-class should lead to execption during SSH-handshakes as you've pointed out in the thread of the issue itself.

Reply:
Yes, but I don't think we should rely on that behavior.

Also, whilst the `KnownHosts` class isn't public, it is a `HostKeyRepository` implementation, which a user can obtain access to via the `JSch.getHostKeyRepository()` or `Session.getHostKeyRepository()`, which they could then utilize standalone for their own purposes, before creating a session that would trigger the SSH-handshake issue.
Therefore, I think it's best that we not hide-away the potential `hmac-sha1` misconfiguration issue inside the `KnownHosts` class, but rather fail immediately when it is encountered.

Reply:
> All the locations in the JSch and Session classes where we instantiate a KnownHosts instance all appear to be in methods that already allow throwing a JSchException.

That's actually not the case:

<pre>
  public HostKeyRepository getHostKeyRepository(){ 
    if(known_hosts==null) known_hosts=new KnownHosts(this);
    return known_hosts; 
  }
</pre>

I can put it into a RuntimeException and throw that instead

Reply:
Oops, you are correct. Throwing a RuntimeException is probably the best choice, thanks!